File: src/riego/control/control_riego_task.py
from src.patrones.observer.observer import Observer
import threading, time
from src.servicios.agua.agua_agotada_exception import AguaAgotadaException  

class ControlRiegoTask(threading.Thread, Observer[float]):
    def __init__(self, sensor_temperatura, sensor_humedad, 
                 plantacion, plantacion_service):
        threading.Thread.__init__(self, daemon=True)
        Observer.__init__(self)
        
        self._sensor_temperatura = sensor_temperatura
        self._sensor_humedad = sensor_humedad
        self._plantacion = plantacion
        self._plantacion_service = plantacion_service
        
        self._ultima_temperatura = 0.0
        self._ultima_humedad = 0.0
        self._detenido = threading.Event()
        
        # Observar sensores
        sensor_temperatura.agregar_observador(self)
        sensor_humedad.agregar_observador(self)
    
    def actualizar(self, evento: float):
        # Actualiza temperatura o humedad
        pass
    
    def run(self):
        while not self._detenido.is_set():
            if self._debe_regar():
                try:
                    self._plantacion_service.regar(self._plantacion)
                except AguaAgotadaException:
                    print("No hay agua suficiente")
            time.sleep(2.5)
    
    def _debe_regar(self) -> bool:
        return (8 <= self._ultima_temperatura <= 15 
                and self._ultima_humedad < 50)

File: src/riego/control/ControlRiegoTask.java
package riego.control;

import entidades.terrenos.Plantacion;
import excepciones.AguaAgotadaException;
import riego.sensores.HumedadReaderTask;
import riego.sensores.TemperaturaReaderTask;
import servicios.terrenos.PlantacionService;

/**
 * Tarea de control de riego automatizado.
 * Monitorea sensores de temperatura y humedad para decidir cuándo regar.
 *
 * REFACTORIZADO: Usa inyección de dependencias.
 */
public class ControlRiegoTask implements Runnable {
    private final TemperaturaReaderTask tempTask;
    private final HumedadReaderTask humTask;
    private Plantacion finca;
    private final PlantacionService plantacionService;

    private volatile boolean ejecutando = true;

    /**
     * Constructor con inyección de dependencias.
     *
     * @param tempTask Tarea de lectura de temperatura
     * @param humTask Tarea de lectura de humedad
     * @param finca Plantación a regar
     * @param plantacionService Servicio de plantación (inyectado)
     */
    public ControlRiegoTask(TemperaturaReaderTask tempTask, HumedadReaderTask humTask,
                           Plantacion finca, PlantacionService plantacionService) {
        this.tempTask = tempTask;
        this.humTask = humTask;
        this.finca = finca;
        this.plantacionService = plantacionService;
    }

    @Override
    public void run() {
        while (ejecutando) {
            try {
                double temp = tempTask.getUltimaTemperatura();
                double hum = humTask.getUltimaHumedad();

                if (!Double.isNaN(temp) && !Double.isNaN(hum)) {
                    if (temp >= 8 && temp <= 15 && hum < 50) {
                        plantacionService.regar(this.finca);
                    }
                }

                Thread.sleep(2500); // frecuencia de control
            } catch (AguaAgotadaException e) {
                // Agua agotada: detener riego y notificar
                System.err.println(e.getFullMessage());
                System.err.println("Sistema de riego detenido automáticamente por falta de agua.");
                ejecutando = false;
                break;
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }
    }

    public void detener() {
        ejecutando = false;
    }
}
